cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java

```diff
@@ -130,93 +124,123 @@ private DefaultDatasetRuntimeContext(AuthorizationEnforcer enforcer, DatasetAcce
     this.accessRecorder = accessRecorder;
     this.principal = principal;
     this.datasetId = datasetId;
+    this.lineageRecorded = new boolean[AccessType.values().length];
+    this.auditRecorded = new boolean[AccessType.values().length];
   }
 
   @Override
-  public void onMethodEntry(@Nullable Class<? extends Annotation> annotation,
-                            Class<? extends Annotation> defaultAnnotation) {
-    final Deque<Class<? extends Annotation>> callStack = this.callStack.get();
-    Class<? extends Annotation> parentAnnotation = callStack.peekLast();
-
-    if (annotation == null) {
-      // If there is no annotation on the method, either use immediate parent one from the call stack
-      // or use the default one if the call stack is empty
-      annotation = (parentAnnotation == null) ? defaultAnnotation : parentAnnotation;
-    }
+  public void onMethodEntry(@Nullable Class<? extends Annotation> annotation) {
+    CallStack callStack = this.callStack.get();
+    AccessInfo accessInfo;
+    AccessType lineageType;
+    AccessType auditType;
+
+    if (annotation != null) {
+      accessInfo = ANNOTATION_TO_ACCESS_INFO.get(annotation);
+      if (accessInfo == null) {
+        // shouldn't happen
+        throw new DataSetException("Unsupported annotation " + annotation + " on dataset " + datasetId);
+      }
 
-    // Do a quick check if the current permission is allowed by the immediate parent if there is one.
-    if (parentAnnotation != null && !isAllowed(parentAnnotation, annotation)) {
-      throw new DataSetException(
-        String.format("Current scope only allows @%s, but the current method requires @%s",
-                      parentAnnotation.getSimpleName(), annotation.getSimpleName()));
-    }
+      // Perform authorization if the method is annotated.
```

Comment:
The `enforcer.enforce` was written in a very efficient way and expected to get called many times. I think it's better to always call enforce in case if user is trying to manipulate the callstack.